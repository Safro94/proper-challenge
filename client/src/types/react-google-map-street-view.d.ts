declare module 'react-google-map-street-view';
