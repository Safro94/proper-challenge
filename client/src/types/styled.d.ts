import 'styled-components';

interface IPalette {
	main: string;
	text: string;
}

interface IBreakpoints {
	mobile: string;
	tablet: string;
	laptop: string;
	desktop: string;
}

declare module 'styled-components' {
	export interface DefaultTheme {
		breakpoints: IBreakpoints;
		palette: {
			common: {
				black: string;
				gray: string;
				white: string;
				danger: string;
			};
			primary: IPalette;
		};
	}
}
