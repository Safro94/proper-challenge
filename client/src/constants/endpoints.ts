export const PROPERTIES_ENDPOINT: string = '/api/properties';
