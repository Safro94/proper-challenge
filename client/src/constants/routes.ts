export const HOME = '/';
export const PROPERTY_DETAIL = '/property/:propertyId';
export const NEW_PROPERTY = '/property/add';
