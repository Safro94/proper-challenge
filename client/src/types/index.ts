export interface IProperty {
	id: number;
	address: string;
	size: string;
	rooms: number;
	utilities: string;
	tenantName: string;
}

export enum ButtonTypes {
	Primary = 'primary',
	Danger = 'danger',
}

export enum RequestStatus {
	Idle = 'idle',
	Pending = 'pending',
	Resolved = 'resolved',
	Rejected = 'rejected',
}
