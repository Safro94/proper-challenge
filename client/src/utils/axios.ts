import axios, { AxiosError, AxiosRequestConfig } from 'axios';
import https from 'https';

const opts: AxiosRequestConfig = {
	httpsAgent: new https.Agent({ rejectUnauthorized: false }),
};

const interceptorError = (error: AxiosError) => Promise.reject(error.response);

const instance = axios.create(opts);

instance.interceptors.response.use(response => response, interceptorError);
instance.defaults.headers.common['x-api-key'] = process.env.REACT_APP_API_KEY;

export default instance;
