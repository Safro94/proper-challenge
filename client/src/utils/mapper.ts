import { IAddress, IDawaResponse } from '../types';

const mapDawaResponse = (data: IDawaResponse[]): IAddress[] => {
	return data.map(item => ({
		text: item.tekst,
		type: item.type,
	}));
};

export { mapDawaResponse };
