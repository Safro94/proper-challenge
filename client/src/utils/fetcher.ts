import { RequestMethods, RequestStatus } from '../types';

import axios from './axios';

interface IFetcher {
	method?: RequestMethods;
	url: string;
	data?: null | any;
	body?: null | any;
	headers?: null | any;
}

interface IResponse {
	data: any;
	status: RequestStatus;
}

const fetcher = async ({
	method = RequestMethods.GET,
	url,
	body = null,
	headers = null,
}: IFetcher): Promise<IResponse> => {
	let data;
	let status = RequestStatus.Pending;

	try {
		const result = await axios[method](
			url,
			JSON.parse(headers),
			JSON.parse(body)
		);
		data = result.data;
		status = RequestStatus.Resolved;
	} catch (error) {
		status = RequestStatus.Rejected;
		throw error;
	}

	return { data, status };
};

export default fetcher;
