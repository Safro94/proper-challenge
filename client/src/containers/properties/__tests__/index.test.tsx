import { render, screen, waitFor } from '../../../utils/test-utils';

import PropertiesContainer from '..';

import fetcher from '../../../utils/fetcher';
import { PROPERTIES_ENDPOINT } from '../../../constants/endpoints';
import { RequestStatus } from '../../../types';

jest.mock('../../../components/loading', () => () => <div>Loading</div>);
jest.mock('../../../utils/fetcher');

const properties = [
	{
		id: 1,
		tenantName: 'Matias',
		rooms: 3,
		size: '40mt2',
		address: 'abc',
		utilities: 'def',
	},
	{
		id: 2,
		tenantName: 'Charles',
		rooms: 3,
		size: '40mt2',
		address: 'abc',
		utilities: 'def',
	},
];

describe('PropertiesContainer', () => {
	it('should render the loading when the request is not resolved yet', async () => {
		(fetcher as jest.Mock).mockImplementation(() =>
			Promise.resolve({ status: RequestStatus.Pending })
		);

		const { container } = render(<PropertiesContainer />);

		await waitFor(() => {
			expect(container.firstChild).toMatchSnapshot();
			expect(screen.getByText(/loading/i)).toBeInTheDocument();
		});
	});

	it('should render correctly when fetcher returns properties', async () => {
		(fetcher as jest.Mock).mockImplementation(() =>
			Promise.resolve({ data: properties, status: RequestStatus.Resolved })
		);

		const { container } = render(<PropertiesContainer />);

		await waitFor(() => {
			expect(fetcher).toHaveBeenCalledTimes(1);
			expect(fetcher).toHaveBeenCalledWith({
				url: `${process.env.REACT_APP_SERVER_URL}${PROPERTIES_ENDPOINT}`,
			});
		});

		await waitFor(() => {
			expect(container.firstChild).toMatchSnapshot();
		});
	});

	it('should render the text when there are no properties', async () => {
		(fetcher as jest.Mock).mockImplementation(() =>
			Promise.resolve({ data: [], status: RequestStatus.Resolved })
		);

		const { container } = render(<PropertiesContainer />);

		await waitFor(() => {
			expect(fetcher).toHaveBeenCalledTimes(1);
			expect(fetcher).toHaveBeenCalledWith({
				url: `${process.env.REACT_APP_SERVER_URL}${PROPERTIES_ENDPOINT}`,
			});
		});

		await waitFor(() => {
			expect(screen.getByText(/noProperties/i)).toBeInTheDocument();
			expect(container.firstChild).toMatchSnapshot();
		});
	});
});
