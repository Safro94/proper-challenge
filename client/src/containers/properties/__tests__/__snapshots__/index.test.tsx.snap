// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PropertiesContainer should render correctly when fetcher returns properties 1`] = `
<div
  class="sc-bdnxRM jOdtAk"
>
  <div
    class="sc-fujyAs ilgEth"
  >
    <a
      class="sc-iCoGMd hkVfKz"
      href="/property/add"
    >
      <svg
        fill="currentColor"
        height="1em"
        stroke="currentColor"
        stroke-width="0"
        viewBox="0 0 512 512"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M363 277h-86v86h-42v-86h-86v-42h86v-86h42v86h86v42z"
        />
        <path
          d="M256 90c44.3 0 86 17.3 117.4 48.6C404.7 170 422 211.7 422 256s-17.3 86-48.6 117.4C342 404.7 300.3 422 256 422c-44.3 0-86-17.3-117.4-48.6C107.3 342 90 300.3 90 256c0-44.3 17.3-86 48.6-117.4C170 107.3 211.7 90 256 90m0-42C141.1 48 48 141.1 48 256s93.1 208 208 208 208-93.1 208-208S370.9 48 256 48z"
        />
      </svg>
      addProperty
    </a>
  </div>
  <div
    class="sc-gtsrHT bCBPuE"
  >
    <div
      class="sc-hKFxyN heRPeh"
    >
      <label
        class="sc-jSFjdj kdvGLc"
        for="downshift-0-input"
        id="downshift-0-label"
      >
        autocompleteText
      </label>
      <div
        aria-expanded="false"
        aria-haspopup="listbox"
        aria-owns="downshift-0-menu"
        class="sc-dlnjwi iecQLF"
        role="combobox"
      >
        <input
          aria-autocomplete="list"
          aria-controls="downshift-0-menu"
          aria-labelledby="downshift-0-label"
          autocomplete="off"
          class="sc-eCApnc bSXRmS"
          id="downshift-0-input"
          placeholder="autocompleteText"
          value=""
        />
      </div>
    </div>
    <ul
      aria-labelledby="downshift-0-label"
      class="sc-bkbkJK fSawmC"
      id="downshift-0-menu"
      role="listbox"
    >
      <li
        aria-selected="false"
        class="sc-dIsUp kHqbmm"
        id="downshift-0-item-0"
        role="option"
      >
        <div
          class="sc-bqGGPW gFXebX"
        >
          <div
            class="sc-hBMUJo PZmcA"
          >
            <img
              alt="Property 0"
              class="sc-fnVZcZ dgIXpp"
              src="house.jpeg"
            />
          </div>
          <div
            class="sc-jrsJWt hqQffW"
          >
            <div
              class="sc-kEqXSa ccymRA"
            >
              <h4
                class="sc-iqAclL"
              >
                tenant
              </h4>
              <h5
                class="sc-crzoAE emQjSB"
              >
                Matias
              </h5>
            </div>
            <div
              class="sc-kEqXSa ccymRA"
            >
              <h4
                class="sc-iqAclL"
              >
                rooms
              </h4>
              <h5
                class="sc-crzoAE emQjSB"
              >
                3
              </h5>
            </div>
            <div
              class="sc-kEqXSa ccymRA"
            >
              <h4
                class="sc-iqAclL"
              >
                size
              </h4>
              <h5
                class="sc-crzoAE emQjSB"
              >
                40mt2
              </h5>
            </div>
            <div
              class="sc-kEqXSa ccymRA"
            >
              <h4
                class="sc-iqAclL"
              >
                address
              </h4>
              <h5
                class="sc-crzoAE emQjSB"
              >
                abc
              </h5>
            </div>
            <div
              class="sc-kEqXSa ccymRA"
            >
              <h4
                class="sc-iqAclL"
              >
                utilities
              </h4>
              <h5
                class="sc-crzoAE emQjSB"
              >
                def
              </h5>
            </div>
          </div>
          <div
            class="sc-ksluID fsnncj"
          >
            <a
              class="sc-fFSPTT liBlhH"
              href="/property/1"
            >
              viewMore
            </a>
            <button
              class="sc-pNWdM bqZpHG"
              type="button"
            >
              deleteProperty
            </button>
          </div>
        </div>
      </li>
      <li
        aria-selected="false"
        class="sc-dIsUp kHqbmm"
        id="downshift-0-item-1"
        role="option"
      >
        <div
          class="sc-bqGGPW gFXebX"
        >
          <div
            class="sc-hBMUJo PZmcA"
          >
            <img
              alt="Property 1"
              class="sc-fnVZcZ dgIXpp"
              src="house.jpeg"
            />
          </div>
          <div
            class="sc-jrsJWt hqQffW"
          >
            <div
              class="sc-kEqXSa ccymRA"
            >
              <h4
                class="sc-iqAclL"
              >
                tenant
              </h4>
              <h5
                class="sc-crzoAE emQjSB"
              >
                Charles
              </h5>
            </div>
            <div
              class="sc-kEqXSa ccymRA"
            >
              <h4
                class="sc-iqAclL"
              >
                rooms
              </h4>
              <h5
                class="sc-crzoAE emQjSB"
              >
                3
              </h5>
            </div>
            <div
              class="sc-kEqXSa ccymRA"
            >
              <h4
                class="sc-iqAclL"
              >
                size
              </h4>
              <h5
                class="sc-crzoAE emQjSB"
              >
                40mt2
              </h5>
            </div>
            <div
              class="sc-kEqXSa ccymRA"
            >
              <h4
                class="sc-iqAclL"
              >
                address
              </h4>
              <h5
                class="sc-crzoAE emQjSB"
              >
                abc
              </h5>
            </div>
            <div
              class="sc-kEqXSa ccymRA"
            >
              <h4
                class="sc-iqAclL"
              >
                utilities
              </h4>
              <h5
                class="sc-crzoAE emQjSB"
              >
                def
              </h5>
            </div>
          </div>
          <div
            class="sc-ksluID fsnncj"
          >
            <a
              class="sc-fFSPTT liBlhH"
              href="/property/2"
            >
              viewMore
            </a>
            <button
              class="sc-pNWdM bqZpHG"
              type="button"
            >
              deleteProperty
            </button>
          </div>
        </div>
      </li>
    </ul>
  </div>
</div>
`;

exports[`PropertiesContainer should render the loading when the request is not resolved yet 1`] = `
<div>
  Loading
</div>
`;

exports[`PropertiesContainer should render the text when there are no properties 1`] = `
<div
  class="sc-bdnxRM jOdtAk"
>
  <div
    class="sc-fujyAs ilgEth"
  >
    <a
      class="sc-iCoGMd hkVfKz"
      href="/property/add"
    >
      <svg
        fill="currentColor"
        height="1em"
        stroke="currentColor"
        stroke-width="0"
        viewBox="0 0 512 512"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M363 277h-86v86h-42v-86h-86v-42h86v-86h42v86h86v42z"
        />
        <path
          d="M256 90c44.3 0 86 17.3 117.4 48.6C404.7 170 422 211.7 422 256s-17.3 86-48.6 117.4C342 404.7 300.3 422 256 422c-44.3 0-86-17.3-117.4-48.6C107.3 342 90 300.3 90 256c0-44.3 17.3-86 48.6-117.4C170 107.3 211.7 90 256 90m0-42C141.1 48 48 141.1 48 256s93.1 208 208 208 208-93.1 208-208S370.9 48 256 48z"
        />
      </svg>
      addProperty
    </a>
  </div>
  <h1
    class="sc-gKAaRy iseYsU"
  >
    noProperties
  </h1>
</div>
`;
